topdown/eval.go

```diff
@@ -1682,30 +1685,65 @@ func (e evalBuiltin) eval(iter unifyIterator) error {
 	numDeclArgs := len(e.bi.Decl.FuncArgs().Args)
 
 	e.e.instr.startTimer(evalOpBuiltinCall)
+	var err error
+	needNormalUnify := true
+	endIndex := len(e.terms) - 1
 
-	err := e.f(e.bctx, operands, func(output *ast.Term) error {
-
+	// We skip evaluation of the builtin entirely if the NDBuiltinCache is
+	// present, and we have a non-deterministic builtin already cached.
+	if e.bi.Nondeterministic && e.bctx.NDBuiltinCache != nil {
```

Comment:
💭 How about adding a helper,
```go
func (e *eval) nbCache(bi *ast.Builtin) bool {
    return bi.Nondeterministic && e.bctx.NDBuiltinCache != nil
}
```
so we could use
```go
if e.e.nbCache(e.bi) {
```
here and in other places?

Reply:
Thanks for the suggestion! I've rolled it into my recent changes. :+1: